cocotb/drivers/amba.py

```diff
@@ -129,86 +251,229 @@ def write(
+                        rresp.append(AXIxRESP(self.bus.RRESP.value.integer))
+                        break
+                    await RisingEdge(self.clock)
+
+                if not hasattr(self.bus, "RLAST") or self.bus.RLAST.value:
+                    break
 
-        if int(result):
-            raise AXIProtocolError("Read address 0x%08x failed with RRESP: %d" %
-                                   (address, int(result)))
+                await RisingEdge(self.clock)
 
-        return data
+            if len(data) != length:
+                raise AXIReadBurstLengthMismatch(
+                    "AXI4 slave returned {} data than expected (requested {} "
+                    "words, received {})"
+                    .format("more" if len(data) > length else "less",
+                            length, len(data)))
+
+            if return_rresp:
+                return list(zip(data, rresp))
+            else:
+                for beat_number, beat_result in enumerate(rresp):
+                    if beat_result is not AXIxRESP.OKAY:
+                        err_msg = "Read on address {0:#x}"
+                        if length != 1:
+                            err_msg += " (beat {1} of {2}, {3} burst)"
+                        err_msg += " failed with RRESP: {4} ({5})"
+
+                        err_msg = err_msg.format(
+                            address, beat_number + 1, length, burst,
+                            beat_result.value, beat_result.name)
+
+                        raise AXIProtocolError(err_msg, beat_result)
+
+                return data
 
     def __len__(self):
         return 2**len(self.bus.ARADDR)
 
 
+class AXI4LiteMaster(AXI4Master):
+    """AXI4-Lite Master"""
+
+    _signals = ["AWVALID", "AWADDR", "AWREADY",        # Write address channel
+                "WVALID", "WREADY", "WDATA", "WSTRB",  # Write data channel
+                "BVALID", "BREADY", "BRESP",           # Write response channel
+                "ARVALID", "ARADDR", "ARREADY",        # Read address channel
+                "RVALID", "RREADY", "RRESP", "RDATA"]  # Read data channel
+
+    _optional_signals = []
+
+    @cocotb.coroutine
+    async def write(self, address, value, byte_enable=None, address_latency=0,
+                    data_latency=0, sync=True):
+        """Write a value to an address.
+
+        Args:
+            address: The address to write to.
+            value: The data value to write.
+            byte_enable: Which bytes in value to actually write. Defaults to
+                None (write all bytes).
+            address_latency: Delay before setting the address (in clock
+                cycles). Default is no delay.
+            data_latency: Delay before setting the data value (in clock
+                cycles). Default is no delay.
+            sync: Wait for rising edge on clock initially. Defaults to True.
+
+            Returns:
+                The write response value.
```

Comment:
I'd like some suggestions here.

Currently, AXI4LiteMaster.write() returns the write response value. However, this does not make sense, as any write response other than `OKAY` just makes it raise an `AXIProtocolError`. In other words, while the docstring promises to return the write response value, in reality it either throws an exception or it returns always the same value.

In my opinion, this method should not return anything.
Is it possibile to deprecate/remove the return value? Or are we forced to add `return AXIxRESP.OKAY` to maintain backwards compatibility?

Reply:
I guess it would be safest to continue returning the response; the overhead is negligible.